src/polyfills/intersection-observer.js

```diff
@@ -19,14 +19,27 @@
  * See https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver.
  */
 
-import {IntersectionObserverStub} from '../polyfillstub/intersection-observer-stub';
+import {
+  IntersectionObserverStub,
+  supportsDocumentRoot,
+} from '../polyfillstub/intersection-observer-stub';
 
 /**
  * @param {!Window} win
  */
 export function install(win) {
-  if (!win.IntersectionObserver) {
+  const NativeIntersectionObserver = win.IntersectionObserver;
+  if (!NativeIntersectionObserver) {
     win.IntersectionObserver = /** @type {typeof IntersectionObserver} */ (IntersectionObserverStub);
+  } else if (!supportsDocumentRoot(win)) {
+    win.IntersectionObserver = function (ioCallback, opts) {
+      if (opts && opts.root && opts.root.nodeType === 9) {
+        return new IntersectionObserverStub(ioCallback, opts);
+      } else {
+        return new NativeIntersectionObserver(ioCallback, opts);
+      }
+    };
+    win.IntersectionObserver.native = NativeIntersectionObserver;
```

Comment:
We probably need to use a quoted prop for this. Overall I rather dislike `native` here and it looks like you too. But I don't have a better suggestion, so perhaps we'll have to live with this? But here's one crazy idea that might be not completely bad. What if we defined our own `opts['forceNative']` prop and your code here would be:

```
if (!(opts && opts['forceNative']) && opts && ....) {
  return new InObStud(...);
} else {
  return new NativeInOb(...);
}
```

Then the relevant branch below can add `'forceNative': true` into options to ensure that it's used here.

Reply:
I'm fine with the idea of the consumer calling new InOb with the custom option, but I don't think that helps us save the NativeInOb cleanly.

The difficult flow is:
1. In startup we synchronously determine that there is an InOb present but it doesn't have doc support
2. We immediately replace the the InOb attached to window with one that uses a `Stub` for document-root InOb and Native for others. Its at this point we need to maintain a reference to the native InOb constructor.
3. Polyfill loads and we attempt to install it. We want to be able to delete the stub and replace it with the polyfill, but only in the cases where we don't prefer native (which is why we still need a reference from step (2))

Reply:
I think we need to first address [this thread](https://github.com/ampproject/amphtml/pull/31453#discussion_r537771502). Depending on how that goes, the points I've made here could become completely moot.